### DISCUSSION TITLE
Renovate is mangling `.npmrc` in subdirectories when post-upgrade tasks exist

### DISCUSSION THREAD

**Discussion by anomiex** [renovatebot/renovate#23489] [Request Help]
### How are you running Renovate?

Self-hosted Renovate

### If you're self-hosting Renovate, tell us what version of Renovate you run.

36.13.2

### If you're self-hosting Renovate, select which platform you are using.

github.com

### Was this something which used to work for you, and then stopped?

It used to work, and then stopped

### Describe the problem

Renovate has started mangling `.npmrc` files in workspaces within the repo when the repo has post-upgrade tasks.

Minimal reproduction repo: https://github.com/anomiex/renovate-test-4/

See https://github.com/anomiex/renovate-test-4/pull/1/files for an example of a renovate PR that wants to change the workspace `.npmrc` files.

The proximate cause of this appears to be #21282, which started passing `{ dot: true }` when processing `fileFilters` after the post-update task. The files had still been mangled previously but were not matched and so did not make it into the commit.

The writes to the files seem to occur at https://github.com/renovatebot/renovate/blob/88452470f992d4f13a68a2595efdada8ca38831d/lib/modules/manager/npm/post-update/index.ts#L160-L168 using data calculated at https://github.com/renovatebot/renovate/blob/88452470f992d4f13a68a2595efdada8ca38831d/lib/modules/manager/npm/extract/index.ts#L104-L142

### Relevant debug logs

<details><summary>Logs</summary>

```
DEBUG: npm file package.json has name undefined (repository=anomiex/renovate-test-4)
DEBUG: npm file packages/a/package.json has name "a" (repository=anomiex/renovate-test-4)
DEBUG: Stripping package-lock setting from .npmrc (repository=anomiex/renovate-test-4)
DEBUG: Package file has no deps (repository=anomiex/renovate-test-4)
DEBUG: npm file packages/b/package.json has name "b" (repository=anomiex/renovate-test-4)
DEBUG: Package file has no deps (repository=anomiex/renovate-test-4)
```
```
DEBUG: Writing package.json files (repository=anomiex/renovate-test-4, branch=renovate-sh/chalk-5.x)
       "packageFiles": ["package.json", "packages/a/package.json", "packages/b/package.json"]
DEBUG: Writing any updated package files (repository=anomiex/renovate-test-4, branch=renovate-sh/chalk-5.x)
DEBUG: Writing package.json (repository=anomiex/renovate-test-4, branch=renovate-sh/chalk-5.x)
DEBUG: npmrc file found in repository (repository=anomiex/renovate-test-4, branch=renovate-sh/chalk-5.x)
DEBUG: Writing updated .npmrc file to .npmrc (repository=anomiex/renovate-test-4, branch=renovate-sh/chalk-5.x)
DEBUG: Generating pnpm-lock.yaml for . (repository=anomiex/renovate-test-4, branch=renovate-sh/chalk-5.x)
DEBUG: Spawning pnpm install to create pnpm-lock.yaml (repository=anomiex/renovate-test-4, branch=renovate-sh/chalk-5.x)
DEBUG: No node constraint found - using latest (repository=anomiex/renovate-test-4, branch=renovate-sh/chalk-5.x)
DEBUG: Setting CONTAINERBASE_CACHE_DIR to /tmp/renovate/cache/containerbase (repository=anomiex/renovate-test-4, branch=renovate-sh/chalk-5.x)
DEBUG: Using containerbase dynamic installs (repository=anomiex/renovate-test-4, branch=renovate-sh/chalk-5.x)
DEBUG: Resolved stable matching version (repository=anomiex/renovate-test-4, branch=renovate-sh/chalk-5.x)
       "toolName": "node",
       "constraint": null,
       "resolvedVersion": "v18.17.0"
DEBUG: Resolved stable matching version (repository=anomiex/renovate-test-4, branch=renovate-sh/chalk-5.x)
       "toolName": "pnpm",
       "constraint": null,
       "resolvedVersion": "8.6.9"
DEBUG: Executing command (repository=anomiex/renovate-test-4, branch=renovate-sh/chalk-5.x)
       "command": "install-tool node v18.17.0"
DEBUG: exec completed (repository=anomiex/renovate-test-4, branch=renovate-sh/chalk-5.x)
       "durationMs": 478,
       "stdout": "[18:39:52.213] INFO (58): Installing tool node v18.17.0...\ntool node v18.17.0 is already installed\ntool is already linked: node v18.17.0\n[18:39:52.499] INFO (58): Installed tool node in 286ms.\n",
       "stderr": ""
DEBUG: Executing command (repository=anomiex/renovate-test-4, branch=renovate-sh/chalk-5.x)
       "command": "install-tool pnpm 8.6.9"
DEBUG: exec completed (repository=anomiex/renovate-test-4, branch=renovate-sh/chalk-5.x)
       "durationMs": 1967,
       "stdout": "[18:39:52.690] INFO (93): Installing tool pnpm v8.6.9...\ninstalling v2 tool pnpm v8.6.9\nlinking tool pnpm v8.6.9\n8.6.9\n[18:39:54.468] INFO (93): Installed tool pnpm in 1.7s.\n",
       "stderr": ""
DEBUG: Executing command (repository=anomiex/renovate-test-4, branch=renovate-sh/chalk-5.x)
       "command": "pnpm install --recursive --lockfile-only --ignore-scripts --ignore-pnpmfile"
DEBUG: exec completed (repository=anomiex/renovate-test-4, branch=renovate-sh/chalk-5.x)
       "durationMs": 1799,
       "stdout": "Scope: all 3 workspace projects\nProgress: resolved 1, reused 0, downloaded 0, added 0\nDone in 873ms\nProgress: resolved 1, reused 0, downloaded 1, added 0, done\n",
       "stderr": ""
DEBUG: pnpm-lock.yaml needs updating (repository=anomiex/renovate-test-4, branch=renovate-sh/chalk-5.x)
DEBUG: Updated 1 lock files (repository=anomiex/renovate-test-4, branch=renovate-sh/chalk-5.x)
       "updatedArtifacts": ["pnpm-lock.yaml"]
DEBUG: Setting CONTAINERBASE_CACHE_DIR to /tmp/renovate/cache/containerbase (repository=anomiex/renovate-test-4, branch=renovate-sh/chalk-5.x)
       "dep": "chalk"
DEBUG: Using containerbase dynamic installs (repository=anomiex/renovate-test-4, branch=renovate-sh/chalk-5.x)
       "dep": "chalk"
DEBUG: Executing command (repository=anomiex/renovate-test-4, branch=renovate-sh/chalk-5.x)
       "dep": "chalk",
       "command": ".github/files/renovate-post-upgrade-run.sh renovate-sh/chalk-5.x"
DEBUG: exec completed (repository=anomiex/renovate-test-4, branch=renovate-sh/chalk-5.x)
       "dep": "chalk",
       "durationMs": 15,
       "stdout": "",
       "stderr": ""
DEBUG: Executed post-upgrade task (repository=anomiex/renovate-test-4, branch=renovate-sh/chalk-5.x)
       "dep": "chalk",
       "cmd": ".github/files/renovate-post-upgrade-run.sh renovate-sh/chalk-5.x",
       "stderr": "",
       "stdout": ""
DEBUG: Post-upgrade file saved (repository=anomiex/renovate-test-4, branch=renovate-sh/chalk-5.x)
       "dep": "chalk",
       "file": "package.json",
       "pattern": "**/*"
DEBUG: Post-upgrade file saved (repository=anomiex/renovate-test-4, branch=renovate-sh/chalk-5.x)
       "dep": "chalk",
       "file": "packages/a/.npmrc",
       "pattern": "**/*"
DEBUG: Post-upgrade file saved (repository=anomiex/renovate-test-4, branch=renovate-sh/chalk-5.x)
       "dep": "chalk",
       "file": "packages/b/.npmrc",
       "pattern": "**/*"
DEBUG: Post-upgrade file saved (repository=anomiex/renovate-test-4, branch=renovate-sh/chalk-5.x)
       "dep": "chalk",
       "file": "pnpm-lock.yaml",
       "pattern": "**/*"
DEBUG: 4 file(s) to commit (repository=anomiex/renovate-test-4, branch=renovate-sh/chalk-5.x)
DEBUG: Preparing files for committing to branch renovate-sh/chalk-5.x (repository=anomiex/renovate-test-4, branch=renovate-sh/chalk-5.x)
DEBUG: Setting git author name: Renovate Bot (repository=anomiex/renovate-test-4, branch=renovate-sh/chalk-5.x)
DEBUG: Setting git author email: bot@renovateapp.com (repository=anomiex/renovate-test-4, branch=renovate-sh/chalk-5.x)
DEBUG: git commit (repository=anomiex/renovate-test-4, branch=renovate-sh/chalk-5.x)
       "deletedFiles": [],
       "ignoredFiles": [],
       "result": {
         "author": null,
         "branch": "renovate-sh/chalk-5.x",
         "commit": "9ec93dec9586ee4c46a5bb0146a743028e58006a",
         "root": false,
         "summary": {"changes": 4, "insertions": 8, "deletions": 6}
       }
```

</details>

### Have you created a minimal reproduction repository?

I have linked to a minimal reproduction in the description above

**Comment by rarkins** [comment]
Created bug issue #23528

**Reply by herndlm** [reply]
sorry, I saw the issue is gone and wanted to ask what the status / solution was?

We have a structure like the following
```
/packages/a/.npmrc
/packages/b/package.json
/packages/b/yarn.lock
```

and without fileFilters we get a MR including the .npmrc. With `**/{package.json,package.lock,yarn.lock}` as fileFilters we get a MR with only changes from package.json though, which is kind of weird.

**Comment by rarkins** [comment]
@herndlm no solution so far. If fileFilters do not work as described then please create a reproduction repo of your own so we can take a look

**Reply by herndlm** [reply]
ok, nevermind, I tried again and it worked now. no clue what the reason was.. xD

**Reply by herndlm** [reply]
saw why it didn't work the first time now. it was an unrelated network error that prevent lock file generation. I can live with the fileFilters configured properly. thx for the quick reply!

**Comment by anomiex** [comment]
> sorry, I saw the issue is gone and wanted to ask what the status / solution was?

@herndlm For now the workaround seems to be to either have your post-upgrade task `git restore` any `.npmrc` files or figure out a set of [minimatch](https://www.npmjs.com/package/minimatch) globs that works for your needs without matching the `.npmrc` files.

As for the deleted issue, I had submitted a PR that would have fixed it, then someone pointed out that it would technically be a breaking change as some people's post-upgrade tasks might be relying on the `.npmrc` still being modified at that point. A [lack of tact](https://www.mit.edu/~jcb/tact.html) on my part led to a personality conflict that ultimately resulted in the current situation, about which I won't say more here in the hope of avoiding a re-escalation.

**Reply by herndlm** [reply]
yeah no worries, thx for the summary! I tried configuring `fileFilters` properly with `**/{package.json,package.lock,yarn.lock}` but, as written in the other comment, it did not commit the yarn.lock anymore for reasons I don't understand. Also tested my glob with https://globster.xyz/?q=**%2F%7Bpackage.json%2Cpackage.lock%2Cyarn.lock%7D&f=%2Fpackages%2Fa%2F.npmrc%2C%2Fpackages%2Fb%2Fpackage.json%2C%2Fpackages%2Fb%2Fyarn.lock which is looking good. But as rarkins mentioned above I need to get that looked at first then I suppose.

**Reply by herndlm** [reply]
ok, nevermind, I tried again and it worked now. no clue what the reason was.. xD

**Comment by philippe-granet** [comment]
Same problem for me, as a workaround, add this at the end of post-upgrade task :
```
git restore '*.npmrc'
```

**Reply by philippe-granet** [reply]
It is not sufficient, as I have `RENOVATE_REUSE_PACKAGE_LOCK=true`, it also remove `package-lock.json`: https://github.com/renovatebot/renovate/blob/88452470f992d4f13a68a2595efdada8ca38831d/lib/modules/manager/npm/post-update/index.ts#L172-L177

**Reply by rarkins** [reply]
Not directly related to this topic - please start a new discussion with own reproduction

**Comment by anomiex** [comment]
This seems to have gotten worse in 17c0041e38ebb22e5a39a2b6f4bd89f617aabd9c. Now it copies `.npmrc` from the root to every workspace without one too.

Workarounds probably need to add something like `git clean -f '*/.npmrc'` now in addition to the `git restore`.

**Reply by alan-agius4** [reply]
We also started seeing this since https://github.com/renovatebot/renovate/commit/17c0041e38ebb22e5a39a2b6f4bd89f617aabd9c. Example: https://github.com/angular/dev-infra/pull/2820/files

**Reply by rarkins** [reply]
The reproduction repository is not reproducing for me. The chalk update PR does not contain changes to `.npmrc`: https://github.com/renovate-reproductions/23489/pull/2/files

Maybe the original problem was fixed, and now there's a similar but different one?

**Reply by rarkins** [reply]
Running against a fork of angular/dev-infra reproduces though: https://github.com/renovate-reproductions/23489b/pull/1/files

**Reply by anomiex** [reply]
Did you run that with the configuration included in the repo, or did you use a configuration that doesn't have a post-upgrade task?

Running it from my original reproduction repo, which I've updated to Renovate 40.19.1 (from 36.107.2) and added a test case for the variation added in 17c0041, I find it reproduces both the original issue and the new one: https://github.com/anomiex/renovate-test-4/pull/1/files

**Reply by rarkins** [reply]
Reverting for now: https://github.com/renovatebot/renovate/pull/36049

**Reply by rarkins** [reply]
You maybe have settings in local env or config.js which are impacting it. Try duplicating it and installing the hosted app so we get a "third" attempt

**Reply by anomiex** [reply]
Can the hosted app even have post-upgrade tasks? The documentation at https://docs.renovatebot.com/configuration-options/#postupgradetasks says

> Post-upgrade tasks can only be used on self-hosted Renovate instances.

**Reply by rarkins** [reply]
It actually has a few which are allowed. But now you made me realize that I probably had your commands blocked when I tried to reproduce

**Reply by rarkins** [reply]
Actually, I don't see any post upgrade tasks in the reproduction repo. @anomiex it doesn't seem like it reproduces for this reason

**Reply by anomiex** [reply]
The post upgrade tasks are configured [here](https://github.com/anomiex/renovate-test-4/blob/a700f4a5a44b915a93f09e8b93a7e0406eac1807/.github/renovate-config.js#L11-L16).
